### ISSUE TITLE
Reconsider the token management story

### ISSUE THREAD

**Issue by emirotin** [resin-io/resin-sdk#284]
So the SDK currently is managing the token. I think it doesn't play well with the UI needs:
- the token is saved to the localStorage as `token`. I can imaging site owners would not be happy to see it, at least let's name it like `resin-auth-token`
- I'd like to manage the token myself, including refreshing it. But that means we need a way to pass the token as an extra param to every request.

What do you think?
I'm a bit confused because the SDK model in general does sound OK, but as I work with it from the UI I feel something is wrong about it.

**Comment by pimterry** [comment]
This makes sense, it's interesting. I think the weird part, for me, is that it's doing this in localStorage. If it was purely in memory, and isolated, that would feel cleaner. I set the token at the beginning, and the sdk keeps using it and effortlessly works (refreshing it internally along the way, but I don't need to know that from the outside).

Does that still not feel right to you? Why do you want to manage refreshing for yourself? I'd find handing the token to every method annoying and unusual, personally.

Do we have any examples we really like of similar SDKs solving similar problems, and how they do this in the browser?

**Comment by emirotin** [comment]
Before you replied I've thought more about it and am completely sure I don't want the SDK to implicitly mess with the side effects like the localStorage.

In-memory is OK
Refreshing on my own is needed for several purposes:
1) what if I don't make requests through the SDK for quite some time? My app will be running an daemon (saga) to refresh it on my own
2) SDK seems to do it before each request through https://github.com/resin-io-modules/resin-request/blob/master/lib/request.coffee. This adds an extra round-trip making the requests much slower
3) The token does also contain the relevant user info, I want to react it the fresh data in my redux
4) I need the token for Angular requests (but I can get it from the SDK so not really a problem).

I would make the following changes:
- add the `autoRefreshToken`, defaults to `!isBrowser`
- add `autoPersistToken`, same thing
- add `persistedTokenName`, defaults to `resin-auth-token`
- (optionally) add `persistToken` and `loadPeristedToken` methods - nice to have, but not needed for the UI
- add `setToken` method that I will call each time I have the fresh token. That simply replaces the in-memory cached token and makes it needless to pass it to each method

**Comment by pimterry** [comment]
> SDK seems to do it before each request 

You mean the `whoami` request to check the token validity? It only does it if your token is more than an hour old (see https://github.com/resin-io-modules/resin-request/blob/master/lib/utils.coffee#L59) so it's this isn't adding many extra requests in practice.

Otherwise though all the above makes sense, thanks, I can see the use cases. It's mostly not browser specific though; extra requests and up to date data and needing the token elsewhere is equally plausible server-side. I do think quite a bit of it is specific to our situation too, mainly because we're using the token outside of the SDK in a complex very Resin-focused UI.

I'm not really convinced this should be the default for everybody else. For users who want to get started using the SDK in the browser, it's much harder and more confusing if you have to do token refreshing yourself, and really nice & easy if it Just Works, and for more users that's all they'll want. Doing it automatically also means you get consistent behaviour between Node and the browser, which is great.

I think we should:
* Change the token name in local storage to `resin-auth-token` (I don't think we should bother making this configurable until somebody asks for that though)
* Add an `autoRefreshToken` option (but turned on by default), which controls whether we do automatic token refreshing, and works in Node and browsers.
* Add an `persistToken` option (but turned on by default), which controls whether we persist the token at all, and works in Node and browsers.
* Potentially add persistToken/loadPeristedToken/setToken later, I don't have strong feelings on these.

How does that sound?

As an aside, I quite want us to get back to normal in the SDK and get v6.0.0 out. I've added the three outstanding issues to a milestone to track this and help chase it, and I think this is now the main thing blocking that release. If we're happy with the above it then only breaking change would be the name though, so we can do that part now, release 6.0.0, and incrementally bring in the rest later.

**Comment by emirotin** [comment]
> It only does it if your token is more than an hour old

Oh, that's good, I was lazy to check, only saw that `tokenRefresh` defaults to true and is only forced to `false` in one place.

> I'm not really convinced this should be the default for everybody else.

Makes sense, agree. And to the bullet points above except this one:
* Potentially add persistToken/loadPeristedToken/setToken later, I don't have strong feelings on these.

If we turn off the token auto-refresh and handle it on our own in the UI we need the way to pass the fresh token to the SDK. If you don't want to pass it to the every request (and you don't) we need a way to `setToken` after we do our own refresh so that the new token is used by the SDK.

And I'm fine with the last point re the release plan (but anyway need all these things for the Redux stuff to work properly).

**Comment by pimterry** [comment]
Ok, great, that all makes sense to me.

I'll do `resin-auth-token` name change shortly, which unblocks 6.0.0 for now. I'll leave the rest of this to you to fit into the redux changes, as and when you need it. Do let me know if you'd like a hand there though.

**Comment by thgreasi** [comment]
@LucianBuzzo is this no longer relevant?
